#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [105]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import datetime as dt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error
from sklearn.metrics import mean_squared_log_error
from sklearn.metrics import median_absolute_error
from sklearn.metrics import explained_variance_score
from sklearn.metrics import max_error
import urllib.request
from PIL import Image

## Data

In [106]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("C:/Users/Rafael Ortega/Proyectos/Machine_Learning/test/train 3.csv", index_col = 0)
test = pd.read_csv("C:/Users/Rafael Ortega/Proyectos/Machine_Learning/test/test 3.csv", index_col = 0)


## Data exploration

In [107]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 912 entries, 0 to 911
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Ram               912 non-null    float64
 1   Weight            912 non-null    float64
 2   Memory            912 non-null    float64
 3   ScreenResolution  912 non-null    float64
 4   Company           912 non-null    float64
 5   TypeName          912 non-null    float64
 6   Cpu               912 non-null    float64
 7   Price_euros       912 non-null    float64
dtypes: float64(8)
memory usage: 64.1 KB


In [108]:
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 391 entries, 0 to 390
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Ram               391 non-null    float64
 1   Weight            391 non-null    float64
 2   Memory            391 non-null    float64
 3   ScreenResolution  391 non-null    float64
 4   Company           391 non-null    float64
 5   TypeName          391 non-null    float64
 6   Cpu               391 non-null    float64
dtypes: float64(7)
memory usage: 24.4 KB


In [109]:
train.head()

Ram    Weight    Memory  ScreenResolution   Company  TypeName  \
0  0.066667  0.437340  0.192063          0.000000  0.100133  0.069022   
1  0.200000  0.138107  0.095238          0.650201  0.296709  0.527398   
2  0.200000  0.590793  0.095238          0.141403  0.100133  0.069022   
3  0.200000  0.145780  0.196825          0.141403  0.266894  0.000000   
4  0.200000  0.539642  0.400000          0.053954  0.264510  0.069022   

        Cpu  Price_euros  
0  0.044073        387.0  
1  0.392871       1379.0  
2  0.219587        854.0  
3  0.392871       1483.0  
4  0.034976        519.0

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [110]:
X = train.drop("Price_euros", axis = 1)
y = train['Price_euros']
X.shape

(912, 7)

In [111]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [112]:
# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

### 3. Asignar el modelo (vacío) a una variable


In [113]:
lm = LinearRegression()

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [114]:
lm.fit(X, y)

LinearRegression()

In [115]:
coeff_df = pd.DataFrame(lm.coef_,
                        X.columns,
                        columns=['Coefficient'])
coeff_df.sort_values('Coefficient', ascending=False)

Coefficient
Ram               1703.929908
Cpu               1314.916528
Company            814.422451
ScreenResolution   607.166272
TypeName           511.233798
Weight              96.740848
Memory            -514.424605

In [116]:
predictions = lm.predict(X)
predictions

array([ 292.97971819, 1788.88559029,  901.46618786, 1134.53807865,
        577.76823063, 1548.42635488,  818.47067486,  888.55114532,
       1561.03106075, 1378.48274286, 1780.01117766,  476.49822579,
       1561.03106075, 1154.45603779,  631.57729057, 1241.61114788,
        906.51099383,  245.98154333, 1520.76248006, 1039.04709935,
       2189.59254937, 1857.49064149, 1183.72063908, 1089.59150313,
       1452.70934267,  389.71648618, 1471.60126042, 1255.3225447 ,
       1003.38356779,  486.01141125, 1253.59061137, 1027.9132422 ,
        876.79281465, 1838.13852422, 1671.9970017 , 2199.2095813 ,
       1008.12239599,  737.26998495,  334.11390678, 1792.38213005,
        409.9411712 , 2525.93105198, 2713.99425182,  323.70336992,
       1061.33514064, 1276.82888132,  423.95991849, 1260.62708793,
        539.25113288,  438.14333724, 1663.7157507 ,  517.90610088,
       1845.90615868,  876.79281465,  991.75487254, 1988.83981749,
        951.56607792, 1459.38965696, 1223.60597313, 1001.93691

In [117]:
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error ,mean_squared_error, r2_score

print("MAE: ", mean_absolute_error(y, predictions))
print("MAPE: ", mean_absolute_percentage_error(y,predictions)) # Es el complemento del R2
print("MSE: ", mean_squared_error(y,predictions))
print("RMSE: ", np.sqrt(mean_squared_error(y,predictions))) #es el mas/menos en la misma unidad que el target
print("R2: ", r2_score(y,predictions))

MAE:  249.2108011680395
MAPE:  0.24339158221441787
MSE:  120874.86319400795
RMSE:  347.6706245773548
R2:  0.7395474221433862


In [118]:
# from sklearn.preprocessing import PolynomialFeatures
# poly_feats = PolynomialFeatures(degree = 3)
# poly_feats.fit(X)
# X_poly = poly_feats.transform(X)
# X_poly




# pol_reg = LinearRegression()
# pol_reg.fit(X, y)




In [119]:
# predictions_submit = pol_reg.predict(X_pred)

# from sklearn.metrics import r2_score
# r2_score(pol_reg.predict(X), y)

# df_preds = pd.DataFrame(predictions_submit, columns = ['predictions'])
# df_preds['real_value'] = pd.Series(y)

# df_preds

-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

In [120]:
# X_pred = pd.read_csv("C:/Users/Rafael Ortega/Proyectos/Machine_Learning/test/test 2.csv", index_col = 0)
X_pred = test
X_pred.head()

Ram    Weight    Memory  ScreenResolution   Company  TypeName  \
0  0.096774  0.139651  0.109155          0.141403  0.305557  0.069022   
1  0.032258  0.177057  0.010563          0.173203  0.021309  0.069022   
2  0.000000  0.177057  0.010563          0.141403  0.021309  0.069022   
3  0.096774  0.488778  0.447183          0.141403  0.296709  0.631573   
4  0.000000  0.102244  0.010563          0.000000  0.259904  0.384296   

        Cpu  
0  0.542987  
1  0.000000  
2  0.000000  
3  0.542987  
4  0.020600

In [121]:
X_pred.shape

(391, 7)

**¿Por qué da error?** 

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [122]:
X_pred.head()

Ram    Weight    Memory  ScreenResolution   Company  TypeName  \
0  0.096774  0.139651  0.109155          0.141403  0.305557  0.069022   
1  0.032258  0.177057  0.010563          0.173203  0.021309  0.069022   
2  0.000000  0.177057  0.010563          0.141403  0.021309  0.069022   
3  0.096774  0.488778  0.447183          0.141403  0.296709  0.631573   
4  0.000000  0.102244  0.010563          0.000000  0.259904  0.384296   

        Cpu  
0  0.542987  
1  0.000000  
2  0.000000  
3  0.542987  
4  0.020600

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [123]:
predictions_submit = lm.predict(X_pred)
predictions_submit

array([1267.32008566,  285.5530783 ,  211.27978716, 1407.59439051,
        500.7698533 ,  716.37354864,  507.6835548 , 1611.98459011,
       2020.39379359, 1912.20173163, 1250.54837619, 1808.43113466,
       1340.19400109, 1170.95395983,  714.07674025, 1250.06587819,
       1493.48648261, 2920.66626991,  839.11608353, 1646.35083414,
       1261.95477145, 1625.43910189,  577.44308908, 1165.89922502,
        396.84797368, 2106.92056642, 1507.23019416, 1434.40649368,
        510.39598165, 1912.20173163, 1639.25042578, 1579.81125325,
       1206.668855  ,  895.90868955, 2349.45662019, 1262.49510573,
       1638.76792779, 2066.32667887,  200.13739026, 2229.64093377,
       2142.75554161, 1348.77024623,  410.37423812, 1737.78937497,
       1256.33835211, 2315.66678086, 1358.80343395, 2111.39317204,
       1269.12454893, 2183.29314823, 1096.22731482,  543.68978639,
       1110.10808619, 1251.03087418,  633.27607891, 1144.9598963 ,
       1346.35775626,  429.23303099, 1238.00342836, 1273.35131

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [124]:
sample = pd.read_csv("data/sample_submission.csv")

In [125]:
sample.head()

laptop_ID  Price_euros
0        750          500
1        726          500
2        633          500
3        363          650
4        319          650

In [126]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [127]:
submission = pd.DataFrame({"laptop_ID": sample['laptop_ID'], "Price_euros": predictions_submit})

In [128]:
submission.head()

laptop_ID  Price_euros
0        750  1267.320086
1        726   285.553078
2        633   211.279787
3        363  1407.594391
4        319   500.769853

In [129]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [130]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.laptop_ID.all() == sample.laptop_ID.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [131]:
chequeator(submission)

You're ready to submit!
